### PR TITLE
updates to address issue 66

### LINKED ISSUE #66: Review comments/questions on IO-MTT v0.1
* In section 5.3, I would suggest that `capabilities.MXL` have the same encoding as `misa.MXL`.
* Section 5.4, in the description of which register fields are used by each operation, omits that the CBQRI-related fields are used by `SET_ENTRY`.
* Why not have the `operand-1.PPN` field start at bit 12, so writing an address to the field requires only masking and not a shift operation?

### PR DESCRIPTION
**Pull request by ved-rivos**
close #66

### CODE REVIEW

File: chapter6.adoc [hunk 1]
```diff
@@ -490,17 +555,39 @@ The `PPN` field programs the PPN of the root page of the MTT during the
...
-acts on the PPN range as specified by the `PPN` and `S` fields. When the `PPNV`
-field is 1, the `S` field sets the address range size for the `MTTINVAL`
-operation. With an `S` field value of 0, the range size is 4 KiB. But, when the
-`S` field has a value of 1, the `MTTINVAL` operation focuses on a NAPOT range.
-This range is decided by the low-order bits of the `PPN` field, going up to the
-first low-order 0 bit (inclusive of this position). If the initial low-order 0
-bit position is denoted as `x`, the size of the range is computed as
-`(1 << (12 + x + 1))`.
+For the `MTTINVAL` operation, the `PPNV` field indicates if the `PPN` field is
+valida and the `IOSDIDV` field indicates if the `IOSDID` field is valid for the
```
Comment (SiFiveHolland) [inline chapter6.adoc:559]: typo:
```suggestion
valid and the `IOSDIDV` field indicates if the `IOSDID` field is valid for the
```

File: chapter6.adoc [hunk 2]
```diff
@@ -490,17 +555,39 @@ The `PPN` field programs the PPN of the root page of the MTT during the
...
+| `PPNV` | `IOSDIDV` | Operation
+|   0    |  0        | Invalidates information cached from any MTT for all
+                       supervisor domain address spaces.
+|   0    |  1        | Invalidates information cached from the MTT for the
+                       adddress space of the supervisor domain identified by
+                       the `IOSDID` operand.
+|   1    |  0        | Invalidates information cached from the MTT for the
+                       address range in the `PPN` operand for all supervisor
+                       domain address spaces.
+|   1    |  0        | Invalidates information cached from the MTT for the
```
Comment (SiFiveHolland) [inline chapter6.adoc:587]: typo:
```suggestion
|   1    |  1        | Invalidates information cached from the MTT for the
```

File: chapter6.adoc [hunk 3]
```diff
@@ -490,17 +555,39 @@ The `PPN` field programs the PPN of the root page of the MTT during the
...
+                       address range in the `PPN` operand for the supervisor
+                       domain address spaces identified by the `IOSDID` operand.
```
Comment (SiFiveHolland) [inline chapter6.adoc:589]:
```suggestion
                       domain address space identified by the `IOSDID` operand.
```

File: chapter6.adoc [hunk 4]
```diff
@@ -67,24 +67,85 @@ response to the address translation request on the device translation completion
...
-to these registers from any of the supervisor domains.
+The I/O MTT checker provides a memory-mapped register programming interface that
+is expected to be associated with the RDSM. The RDSM may employ the MTT (or PMP)
+to prohibit access to these registers from any of the supervisor domains.
+
+If an MTT check disallows a transaction then the transaction is aborted.
+
+If the aborted transaction is an IOMMU-initiated memory access then the IO bridge
+signals such access faults to the IOMMU itself. The details of such signaling is
+implementation defined.
```
Comment (SiFiveHolland) [inline chapter6.adoc:119]: On a related note: is it expected that IOMMU-initiated transactions (i.e. page table walks) go through the MTTCHK? If so, we should specify this.

**Review by SiFiveHolland — COMMENTED** [review]
Thanks, this is much clearer. LGTM with a few fixes.

One general editorial comment: this chapter isn't consistent between using "MTTCHK" and "IOMTTCHK".

File: chapter6.adoc [hunk 5]
```diff
@@ -67,24 +67,85 @@ response to the address translation request on the device translation completion
...
+
+* It is generally unsafe to update fields of an MTT entry using stores of width
+  less than the minimal single-copy atomic memory access supported by MTTCHK as
+  it is legal for MTTCHK to read the entry at any time, including when only
+  some of the partial stores have taken effect. For an update to be atomic,
+  software must use a single store of width equal to the minimal single-copy
+  atomic memory access supported by MTTCHK.
+
+* MTTCHK is not required to immediately observe the software updates to an MTT
+  entry. Software must use the `MTTINVAL` operation outlined in <<section>> to
```
Comment (rsahita) [inline chapter6.adoc:99]: section name to be filled

File: chapter6.adoc [hunk 6]
```diff
@@ -471,15 +535,16 @@ operations requested through `control.OP`.
   {bits:  4, name: 'MTT_MODE (WARL)'},
   {bits:  1, name: 'PPNV (WARL)'},
   {bits:  1, name: 'S (WARL)'},
+  {bits:  1, name: 'RV'},
```
Comment (SiFiveHolland) [inline chapter6.adoc:538]:
```suggestion
  {bits:  1, name: 'IOSDIDV'},
```

File: chapter6.adoc [hunk 7]
```diff
@@ -67,24 +67,85 @@ response to the address translation request on the device translation completion
...
+`MTTINVAL` operation to synchronize updates to the entry with the operation of
+the MTTCHK. These are the only behaviors expected.
+====
 
-The I/O MTT checker provides a memory-mapped register programming interface
-associated with the RDSM. The RDSM employs the MTT (or PMP) to prohibit access
-to these registers from any of the supervisor domains.
+The I/O MTT checker provides a memory-mapped register programming interface that
+is expected to be associated with the RDSM. The RDSM may employ the MTT (or PMP)
+to prohibit access to these registers from any of the supervisor domains.
```
Comment (rsahita) [inline chapter6.adoc:113]: "expected to be associated with" -> expected to be M-mode accessible?

Do we need to say "The RDSM may employ..." - like the other case that is implementation choice (probably a sane one) but no need to specify it?

File: chapter6.adoc [hunk 8]
```diff
@@ -521,3 +608,18 @@ of the `PPN` field.
 |===
 
 ====
+
+[NOTE]
+====
+Simpler implementations may ignore the operands of `MTTINVAL` operation and
+perform a global invalidation of all inforamtion cached from any MTT.
```
Comment (rsahita) [inline chapter6.adoc:615]:
```suggestion
perform a global invalidation of all information cached from any MTT.
```

File: chapter6.adoc (hunk 2 above)
Comment (rsahita) [inline chapter6.adoc:587]:
```suggestion
|   1    |  1        | Invalidates information cached from the MTT for the
```

**Review by rsahita — CHANGES_REQUESTED** [review]
some minor changes proposed

File: chapter6.adoc (hunk 4 above)
Comment (ved-rivos) [reply chapter6.adoc:119]: It's stated here:
> The MTTCHK uses the SDID to determine the MTT associated with the supervisor domain and checks if the physical address may be accessed by the device or IOMMU associated with that supervisor domain.

File: chapter6.adoc (hunk 7 above)
Comment (ved-rivos) [reply chapter6.adoc:113]: Yes, its repeating the obvious. So I will remove the "that is expected ...." and the sentence that follows it.

**Comment by ved-rivos** [comment]
> Thanks, this is much clearer. LGTM with a few fixes.
> 
> One general editorial comment: this chapter isn't consistent between using "MTTCHK" and "IOMTTCHK".

IOMTTCHK include SDCL and MTTCHK. The chapter refers broadly to IOMTTCHK when appropriate and to MTTCHK when MTT checking is discussed.

**Review by rsahita — APPROVED** [review]
lgtm

**Pull request merged**